### PR TITLE
feat: Skip RR Assignment if Contact Exists In Salesforce

### PR DESCRIPTION
**Pull request by joeauyeung**
## What does this PR do?

- In Salesforce when creating new contacts, assign the owner to the booking organizer if they exist
	- Fallback to the credential owner account
- In a round robin event, if the email param is present in the URL, skip round robin assignment and create the booking under the CRM contact owner

Fixes CAL-3370

https://www.loom.com/share/2a3d126c82d94997ab957469536e2930?sid=ebeaf6a3-2e2c-41e7-8519-6a8c4ea4f312

## Requirement/Documentation

- If there is a requirement document, please, share it here.
- If there is a UI/UX design document, please, share it here.

## Type of change

- New feature (non-breaking change which adds functionality)

## How should this be tested?

- Connect a Salesforce account
- Create a round robin event
- Enable the Salesforce app under the round robin event and enable skip round robin  assignment
- Create a booking under the rr event
	- Note the owner of the new contact is the team member that's assigned if they exist in Salesforce
- Create another booking under the rr event. This time add the email of the attendee of the previous booking as a URL param
	- Availability should only be pulled for the contact owner and the new booking should be assigned to them

## Mandatory Tasks

- [ ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/app-store/salesforce/lib/CrmService.ts [hunk 1]
```diff
@@ -113,37 +112,14 @@ export default class SalesforceCRMService implements CRM {
     });
   };
 
-  private salesforceContactCreate = async (attendees: Person[]) => {
```
Comment (joeauyeung) [inline packages/app-store/salesforce/lib/CrmService.ts:116]: Wasn't being used anywhere

File: packages/app-store/salesforce/lib/CrmService.ts [hunk 2]
```diff
@@ -113,37 +112,14 @@ export default class SalesforceCRMService implements CRM {
...
+  private getSalesforceUserFromEmail = async (email: string) => {
```
Comment (joeauyeung) [inline packages/app-store/salesforce/lib/CrmService.ts:115]: Get the Salesfoce user info from the email provided. Assuming that the email on Cal.com is the same as Salesforce

File: packages/app-store/salesforce/lib/CrmService.ts [hunk 3]
```diff
@@ -113,37 +112,14 @@ export default class SalesforceCRMService implements CRM {
...
-      })
-    );
-    return createdContacts.filter(
-      (contact): contact is Omit<ContactSearchResult, "attributes"> => contact !== undefined
-    );
+    return await conn.query(`SELECT Id, Email FROM User WHERE Email = '${email}'`);
   };
 
-  private salesforceContactSearch = async (event: CalendarEvent) => {
+  private getSalesforceUserFromOwnerId = async (ownerId: string) => {
```
Comment (joeauyeung) [inline packages/app-store/salesforce/lib/CrmService.ts:120]: When querying contacts, it only returns the `ownerId`. So need to make another query to Salesfoce for the user email.

File: packages/app-store/salesforce/lib/CrmService.ts [hunk 4]
```diff
@@ -276,11 +252,34 @@ export default class SalesforceCRMService implements CRM {
...
     const conn = await this.conn;
     const emails = Array.isArray(email) ? email : [email];
-    const soql = `SELECT Id, Email FROM Contact WHERE Email IN ('${emails.join("','")}')`;
+    // TODO get the user associated with the contact. Return the email
+    const soql = `SELECT Id, Email, OwnerId FROM Contact WHERE Email IN ('${emails.join("','")}')`;
     const results = await conn.query(soql);
+
+    if (!results || !results.records.length) return [];
+
+    if (includeOwner) {
```
Comment (joeauyeung) [inline packages/app-store/salesforce/lib/CrmService.ts:264]: We have to make different calls to Salesfoce for the owner information. Let's hide this logic behind a param to only make those calls if we need to.

File: packages/app-store/salesforce/lib/CrmService.ts [hunk 5]
```diff
@@ -289,8 +288,17 @@ export default class SalesforceCRMService implements CRM {
...
     const conn = await this.conn;
+
+    // See if the organizer exists in the CRM
+    let organizerId: string;
+    if (organizerEmail) {
+      const userQuery = await this.getSalesforceUserFromEmail(organizerEmail);
+      if (userQuery) {
+        organizerId = (userQuery.records[0] as { Email: string; Id: string }).Id;
+      }
+    }
```
Comment (joeauyeung) [inline packages/app-store/salesforce/lib/CrmService.ts:301]: See if the organizing user exists in Salesforce. If so then connect the new contact to that organizer. If they don't exist then fallback to the account that iniated the Salesforce OAuth.

File: packages/core/crmManager/crmManager.ts
Comment (joeauyeung) [inline packages/core/crmManager/crmManager.ts:1]: Pass the organizer email to the create contact method to connect the new contacts to the right owner.

File: packages/features/bookings/Booker/components/hooks/useBookings.ts [hunk 6]
```diff
@@ -25,6 +25,7 @@ export interface IUseBookings {
   hashedLink?: string | null;
   bookingForm: UseBookingFormReturnType["bookingForm"];
   metadata: Record<string, string>;
+  teamMemberEmail?: string;
```
Comment (joeauyeung) [inline packages/features/bookings/Booker/components/hooks/useBookings.ts:28]: This is the email of the team member that is the owner of the contact in Salesforce. Pass it to `handleNewBooking` via this variable.

File: packages/features/bookings/Booker/utils/event.ts [hunk 7]
```diff
@@ -60,6 +60,7 @@ export const useScheduleForEvent = ({
   dayCount,
   selectedDate,
   orgSlug,
+  bookerEmail,
```
Comment (joeauyeung) [inline packages/features/bookings/Booker/utils/event.ts:63]: This is the email that is taken from the URL param

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 8]
```diff
@@ -1245,6 +1245,13 @@ async function handler(
           luckyUserPool: luckyUserPool.map((user) => user.id),
         })
       );
+
+      if (reqBody.teamMemberEmail) {
+        const teamMember = eventTypeWithUsers.users.find((user) => user.email === reqBody.teamMemberEmail);
+        if (teamMember) {
+          luckyUsers.push(teamMember);
+        }
+      }
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:1254]: If the `teamMemberEmail` is passed then let's skip the logic of finding the lucky user.

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 9]
```diff
@@ -1298,7 +1305,10 @@ async function handler(
         throw new Error(ErrorCode.HostsUnavailableForBooking);
       }
       // Pushing fixed user before the luckyUser guarantees the (first) fixed user as the organizer.
-      users = [...availableUsers.filter((user) => user.isFixed), ...luckyUsers];
+      users = [
+        ...availableUsers.filter((user) => user.isFixed && user.email !== reqBody.teamMemberEmail),
+        ...luckyUsers,
+      ];
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:1311]: If the team member is a fixed host then remove them from this array so they aren't double booked.

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 10]
```diff
@@ -344,13 +345,69 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions) {
...
+          user: {
+            select: {
+              email: true,
+            },
+          },
+        },
+      });
+      if (crmCredential) {
+        const crm = new CrmManager(crmCredential);
+        const contact = await crm.getContacts(input.bookerEmail, true);
```
Comment (joeauyeung) [inline packages/trpc/server/routers/viewer/slots/util.ts:383]: I was scared about making one more call when loading slots but if we narrow down the host we only need to get slots for that one host vs all hosts under an event type.

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 11]
```diff
@@ -655,6 +712,7 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions) {
 
   return {
     slots: computedAvailableSlots,
+    teamMember,
```
Comment (joeauyeung) [inline packages/trpc/server/routers/viewer/slots/util.ts:715]: Return the team member email to eventually send to `handleNewBooking`

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 12]
```diff
@@ -344,13 +345,69 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions) {
...
+        const contact = await crm.getContacts(input.bookerEmail, true);
+        if (contact) {
+          // Since this is enabled for round robin event, we iterate through hosts
+          const contactOwner = eventType.hosts.find((host) => host.user.email === contact[0].ownerEmail);
+          if (contactOwner) {
+            teamMember = contactOwner.user.email;
+            usersWithCredentials = [
+              contactOwner.user,
+              ...eventType.hosts.reduce((fixedHosts, host) => {
+                if (host.user.email !== contactOwner.user.email && host.isFixed) fixedHosts.push(host.user);
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/slots/util.ts:392]: What if the contactOwner is a fixed host. Then we still want to add the round robin hosts here, no?

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
As discussed, this should only handle round robin hosts. If the contact owner is a fixed host they will anyways be part of the booking

File: packages/trpc/server/routers/viewer/slots/util.ts (hunk 11 above)
Comment (CarinaWolli) [reply packages/trpc/server/routers/viewer/slots/util.ts:715]: can we make that name clearer, something like `crmContactOwner`?

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 13]
```diff
@@ -1260,7 +1260,12 @@ async function handler(
         loggerWithEventDetails
       );
       const luckyUsers: typeof users = [];
-      const luckyUserPool = availableUsers.filter((user) => !user.isFixed);
+      const luckyUserPool: IsFixedAwareUser[] = [];
+      const fixedUserPool: IsFixedAwareUser[] = [];
+      availableUsers.forEach((user) => {
+        user.isFixed ? fixedUserPool.push(user) : luckyUserPool.push(user);
+      });
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:1267]: We iterate through the `availableUsers` array multiple times here so let's just do it once.

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 14]
```diff
@@ -1272,11 +1277,18 @@ async function handler(
...
       if (reqBody.teamMemberEmail) {
-        const teamMember = eventTypeWithUsers.users.find((user) => user.email === reqBody.teamMemberEmail);
-        if (teamMember) {
-          luckyUsers.push(teamMember);
+        // If requested user is not a fixed host, assign the lucky user as the team member
+        if (!fixedUserPool.some((user) => user.email === reqBody.teamMemberEmail)) {
+          const teamMember = eventTypeWithUsers.users.find((user) => user.email === reqBody.teamMemberEmail);
+          if (teamMember) {
+            luckyUsers.push(teamMember);
+          }
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:1285]: @CarinaWolli if the contact owner is a fixed host then let's still find a lucky user

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 15]
```diff
@@ -1324,16 +1336,9 @@ async function handler(
         }
       }
       // ALL fixed users must be available
-      if (
-        availableUsers.filter((user) => user.isFixed).length !== users.filter((user) => user.isFixed).length
-      ) {
+      if (fixedUserPool.length !== users.filter((user) => user.isFixed).length) {
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:1339]: Removing an iteration of `availableUsers`

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 16]
```diff
@@ -1350,7 +1355,10 @@ async function handler(
     throw new Error(ErrorCode.NoAvailableUsersFound);
   }
 
-  const [organizerUser] = users;
+  // If the team member is requested then they should be the organizer
+  const organizerUser = reqBody.teamMemberEmail
+    ? users.find((user) => user.email === reqBody.teamMemberEmail)
+    : users[0];
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:1361]: Better way to ensure the requested team member is the organizer

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 17]
```diff
@@ -362,13 +363,69 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Pro
...
+            select: {
+              email: true,
+            },
+          },
+        },
+      });
+      if (crmCredential) {
+        const crm = new CrmManager(crmCredential);
+        const contact = await crm.getContacts(input.bookerEmail, true);
+        if (contact) {
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/slots/util.ts:402]: this is an array, we need to check if `length > 0 `

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 18]
```diff
@@ -362,13 +363,69 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Pro
...
+          // Since this is enabled for round robin event, we iterate through hosts
+          const contactOwner = eventType.hosts.find((host) => host.user.email === contact[0].ownerEmail);
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/slots/util.ts:404]: `contact[0].ownerEmail` does this work for you? it only returns contact id and email for me but not the contact owner email

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
Left some blocking comments that need to be addressed 🙏

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 19]
```diff
@@ -362,13 +363,69 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Pro
...
+  let teamMember;
+
+  if (eventType.schedulingType === SchedulingType.ROUND_ROBIN && input.bookerEmail) {
+    let crmRoundRobinLeadSkip;
+    // See if CRM app is enabled and skip RR assignment
+    const eventTypeAppMetadata = eventType?.metadata?.apps;
+    for (const appKey in eventTypeAppMetadata) {
+      const app = eventTypeAppMetadata[appKey];
+      if (
+        app.enabled &&
```
[the remark this answers is not in the document]
Comment (joeauyeung) [reply packages/trpc/server/routers/viewer/slots/util.ts:376]: I found the enabled status for the event type apps was flakey. What do you think of this solution?
[`7dad5c3` (#14556)](https://github.com/calcom/cal.com/pull/14556/commits/7dad5c39f31a868ffd8a819e2e6dec6cc77e5a1e)

File: packages/trpc/server/routers/viewer/slots/util.ts (hunk 18 above)
Comment (joeauyeung) [reply packages/trpc/server/routers/viewer/slots/util.ts:404]: Forgot to pass the param to include the contact host. Should be good now.
[`2c05c8e` (#14556)](https://github.com/calcom/cal.com/pull/14556/commits/2c05c8e3474c42c65ad2b5721fbde4a1d6116966)

File: packages/trpc/server/routers/viewer/slots/util.ts (hunk 12 above)
Comment (joeauyeung) [reply packages/trpc/server/routers/viewer/slots/util.ts]: Good point. How do we feel about assigning the contact owner as a fixed host so the slots available are always available for the contact owner?
[`7bca279` (#14556)](https://github.com/calcom/cal.com/pull/14556/commits/7bca2794622d4e86f3fa906b4b5bfbfc0f3a1cd5)

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 20]
```diff
@@ -362,13 +363,69 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Pro
...
+            select: {
+              email: true,
+            },
+          },
+        },
+      });
+      if (crmCredential) {
+        const crm = new CrmManager(crmCredential);
+        const contact = await crm.getContacts(input.bookerEmail, true);
+        if (contact.length > 0) {
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/slots/util.ts:402]:
```suggestion
        if (contact?.length) {
```
type error

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 21]
```diff
@@ -362,13 +363,69 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Pro
...
+          // Since this is enabled for round robin event, we iterate through hosts
+          const contactOwner = eventType.hosts.find((host) => host.user.email === contact[0].ownerEmail);
+          if (contactOwner) {
+            teamMember = contactOwner.user.email;
+            usersWithCredentials = [
+              { ...contactOwner.user, isFixed: true },
+              ...eventType.hosts.reduce((hostsArray, host) => {
+                if (host.user.email !== contactOwner.user.email) hostsArray.push(host.user);
+                return fixedHosts;
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/slots/util.ts:411]: `fixedHosts` is not defined anywhere

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 22]
```diff
@@ -362,13 +363,69 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Pro
...
+        const contact = await crm.getContacts(input.bookerEmail, true);
+        if (contact.length > 0) {
+          // Since this is enabled for round robin event, we iterate through hosts
+          const contactOwner = eventType.hosts.find((host) => host.user.email === contact[0].ownerEmail);
+          if (contactOwner) {
+            teamMember = contactOwner.user.email;
+            usersWithCredentials = [
+              { ...contactOwner.user, isFixed: true },
+              ...eventType.hosts.reduce((hostsArray, host) => {
+                if (host.user.email !== contactOwner.user.email) hostsArray.push(host.user);
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/slots/util.ts:410]: `hostsArray.push(host.user)` has a type error: 
```
Argument of type '{ availability: { id: number; userId: number | null; eventTypeId: number | null; startTime: Date; endTime: Date; date: Date | null; scheduleId: number | null; days: number[]; }[]; ... 12 more ...; selectedCalendars: { ...; }[]; }' is not assignable to parameter of type 'never'.
```

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 23]
```diff
@@ -1319,21 +1335,18 @@ async function handler(
...
+      users = [...fixedUserPool, ...luckyUsers];
       luckyUserResponse = { luckyUsers: luckyUsers.map((u) => u.id) };
     } else if (req.body.allRecurringDates && eventType.schedulingType === SchedulingType.ROUND_ROBIN) {
       // all recurring slots except the first one
       const luckyUsersFromFirstBooking = luckyUsers
         ? eventTypeWithUsers.users.filter((user) => luckyUsers.find((luckyUserId) => luckyUserId === user.id))
         : [];
-      const fixedHosts = eventTypeWithUsers.users.filter((user: IsFixedAwareUser) => user.isFixed);
-      users = [...fixedHosts, ...luckyUsersFromFirstBooking];
+      users = [...fixedUserPool, ...luckyUsersFromFirstBooking];
```
Comment (CarinaWolli) [inline packages/features/bookings/lib/handleNewBooking.ts:1349]: `fixedUserPool` can't be accessed in the else part

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 24]
```diff
@@ -1342,7 +1355,10 @@ async function handler(
...
   }
 
-  const [organizerUser] = users;
+  // If the team member is requested then they should be the organizer
+  const organizerUser = reqBody.teamMemberEmail
+    ? users.find((user) => user.email === reqBody.teamMemberEmail)
+    : users[0];
 
   const tOrganizer = await getTranslation(organizerUser?.locale ?? "en", "common");
   const allCredentials = await getAllCredentials(organizerUser, eventType);
```
Comment (CarinaWolli) [inline packages/features/bookings/lib/handleNewBooking.ts:1364]: Type error for organizerUser: 

```
Type 'undefined' is not assignable to type '{ id: number; username: string | null; credentials: { id: number; type: string; user: { email: string; } | null; userId: number | null; teamId: number | null; key: JsonValue; appId: string | null; invalid: boolean | null; }[]; }'.
```

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 25]
```diff
@@ -1342,7 +1355,10 @@ async function handler(
     throw new Error(ErrorCode.NoAvailableUsersFound);
   }
 
-  const [organizerUser] = users;
+  // If the team member is requested then they should be the organizer
+  const organizerUser = reqBody.teamMemberEmail
```
Comment (CarinaWolli) [inline packages/features/bookings/lib/handleNewBooking.ts:1359]: `organizerUser` causes some type error below because it can be `undefined` now

**Review by CarinaWolli — CHANGES_REQUESTED** [review]

**Review by CarinaWolli — DISMISSED** [review]
Found a bug 

Settings: 
- teampro is contact owner of test3@test.com
Screenshot 2024-05-15 at 11 54 30 AM


Fixed host was not assigned but teampro as assigned twice:
Screenshot 2024-05-15 at 11 51 54 AM

**Review by zomars — APPROVED** [review]
Seeing from the back and forth. Code looks fine so far and tests are passing as well. I do have some observations tho:
- We got a huge amount of nested if/else statements, there a lot of conditionals that could easily be a single-purpose function.
- We're lacking both unit and integration tests to prove that this works and won't break in future updates.

Since this is not being merged to `main` yet and it is going to the `crm-manager` branch. I'll approve and merge this but in order to get `crm-manager` into main we need to address those concerns.

Nice work @joeauyeung and @CarinaWolli with the development and thorough testing so far. 🙏

**Pull request merged**